File: group-anagrams/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        def prime_hash(str):
            prime_factors = [
                2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 
                31, 37, 41, 43, 47, 53, 59, 61, 67, 
                71, 73, 79, 83, 89, 97, 101
            ] 
            hash = 1
            for i in str:
                hash *= prime_factors[(ord(i) - 97)]
            
            return hash

        hashes = []
        for i in strs:
            hashes.append(prime_hash(i))
        
        coupled = [(i,j) for i,j in zip(hashes, strs)]

        coupled = sorted(coupled, key=lambda x: x[0])

        answer = []

        temp = [coupled[0]]
        n = 0 
        for i in coupled[1:]:
            if i[0] == temp[0][0]:
                temp.append(i)
            else:
                answer.append(temp)
                temp = [i]
        answer.append(temp)
        
        answer = [[j[1] for j in i] for i in answer]

        return answer

File: group-anagrams/group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> dict;
        string hash;
        for (auto& s : strs){
            dict[optimised_sort(s)].push_back(s);  
        }

        vector<vector<string>> answer;

        for (auto& i : dict){
            answer.push_back(i.second);
        }

        return answer;
    }

    string optimised_sort(string s){
        int freq[26] = {0};
        for (auto& c:s){
            freq[c -'a']++;
        }
        string sorted;
        for (int i = 0; i < 26; i++){
            sorted += string(freq[i], i + 'a');
        }
        return sorted;
    }
};
